### DISCUSSION TITLE
Creating your own chat history for API?

### DISCUSSION THREAD

**Discussion by twelchdev** [oobabooga/text-generation-webui#1743] [Q&A]
So been working with the API and most things are working great!  But because it is stateless I was wondering if there was a way I could provide a chat history to the api? I can easily store past prompts in json or database. But what would be a good way to provide these to in an api request?  

Is there a request parameter for this?
Would you just add it to your prompt? Could get huge but you could limit the history you provide to a length where you won't get errors?

Sorry if I missed some documentation on how to use chat history, but it seems with the api there isn't much guidance. Anyhow just curious if anyone has ideas or has tried this out?

Thanks!

**Comment by LaaZa** [comment]
Handle the history in your application and send API the prompt you can fit. You generally want to have some context that will always stay in the prompt that establishes the chat, like small guidance for the assistant or a character it should play. 
The API also has `/api/v1/token-count` endpoint where you can send the prompt to count tokens with the correct tokenizer without loading it in your code.

Remember to use the correct pattern for the specific model when constructing the prompt for best results.
For more advanced history and prompt management you can consider different methods of dynamically fetching related history like using a semantic search. Or generating a summary to condence past history to main points. Since the AI is stateless you can easily use a completely new prompt to generate summary or similar if it is capable of that.

**Reply by TomLucidor** [reply]
Probably need a checklist equivalent of this advice since "Role Playing" and working autonomous agents have different goals but the same need for memory.

**Comment by twelchdev** [comment]
So basically I can make the request as large as my machine can handle without getting CUDA errors.  And I can use` /api/vi/token-count` to count tokens before sending the request to check tokens and prevent getting any errors?

So then I just have to construct the right pattern for the model which includes some context about chat history?  I like the idea of generating a condensed history. But If I wanted to just use the literal input/response history and I am using alpaca, because vicuna seems not to work with api **(I believe you need Vicuna0 instruction template to prevent bug where Vicuna prompts itself repeatedly when using instruct**). Where would I add conversation history? At the top as part of the context? The format for alpaca-with-input is:

```
Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Instruction

### Input:
Input

### Response:
```
So as a silly example, something like

```
Below is an instruction that describes a task, I want you to respond to the task as if you are a Doctor named 'House'. Dr House is a genius doctor who specializes in diagnosing uncured diseases. He is gruff and sarcastic but always works hard to get to the bottom of his medical case. Here is Dr. House's most recent conversation with his patient.
Patient: Hello Doctor, I am not feeling so good today. Can you help me?
Dr. House: Of course! What's causing you discomfort?

### Instruction:
Respond to the following as if you are Dr. House taking your previous conversation into account.

### Input:
I have Muscle and joint pain. And I am experiencing pain and stiffness

### Response:
```
I tried this out through web-ui and alpaca seems to pretend like there has been a previous conversation, but is making up extra details. This is the response I get:
_As I mentioned before, it could be Lupus or Rheumatoid Arthritis. However, we need more tests to confirm the diagnosis. In the meantime, I would recommend resting and taking over-the-counter pain relievers. Please make sure to drink plenty of water and maintain a healthy diet. We will continue to investigate the cause and find the right treatment for you._

**Reply by twelchdev** [reply]
found Vicuna0 instruction template. Going to try it out:
```
name: "### Assistant:"
your_name: "### Human:"
context: "A chat between a human and an assistant.\n\n"
turn_template: "<|user|> <|user-message|>\n<|bot|> <|bot-message|>\n"
```

**Reply by Claudioappassionato** [reply]
it worked? I am very interested in having something remembered, or in recovering traces of conversations. For example, writing a book is impossible. Another question in the WebUi console is there a Chat history to load. I've always had an error and I don't understand its function. Claudio

**Comment by w0lfzk1n** [comment]
Hello there. So i think this could might help you, understand how to use the history thing in oobabooga.
It took me a while to figure it out for a kind of chatbot, but here you go.
This example is in javascript, but you should get the hang of it, i hope:

==============================================================================

```
const axios = require('axios');
const readline = require('readline');

const HOST = 'localhost:5000';
const URI = `http://${HOST}/api/v1/chat`;

// Important thing for the AI to remember and use the history correctly!
// The Language Model will always work on the second string, in the last object of the history.internal or *.visible object.
// Therefore, when a new "human" prompt is entered, before making the API request, you need to push the user_input
// and a empty string (or the string containing the start of the AI's reply) which the AI will then use for the generation of it's prompt.

// It took me some time until i have figured it out, since there isn't much information about it's actual usage on the web.
// Or i wasn't able to find anything, so I thought i share my knowledge, maybe helping out some of you, struggling with this.

/*EXAMPLE HISTORY WITH A NEW USERINPUT BEFORE SENDING WITH API 

'''
let history = {
    internal: [ ["Hi there","Hello, I am an AI. How can i help you?"], ["Please explain how language models work",""] ],
    visible: [ ["Hi there","Hello, I am an AI. How can i help you?"], ["Please explain how language models work",""] ]
}
'''

Optional you can here for example use "Language models are" for giving a emtpy "" string for the AI to work on.

*/

let history = {
    internal: [],
    visible: []
}

// API Request function
async function run(user_input, history) {
    const request = {
        user_input: user_input,
        history: history,
        mode: 'chat',  // Options: 'chat', 'chat-instruct', 'instruct'
        character: "Example",
        instruction_template: 'vicuna-v1.1',
        your_name: 'Human',

        regenerate: false,
        _continue: true, // Option for sending History
        stop_at_newline: false,
        chat_prompt_size: 2048,
        chat_generation_attempts: 1,
        chat_instruct_command: 'Continue the chat dialogue below. Write a single reply for the character "Assistant"\n\n',

        max_new_tokens: 250,
        do_sample: true,
        temperature: 0.7,
        top_p: 0.1,
        typical_p: 1,
        epsilon_cutoff: 0,
        eta_cutoff: 0,
        tfs: 1,
        top_a: 0,
        repetition_penalty: 1.18,
        top_k: 40,
        min_length: 0,
        no_repeat_ngram_size: 0,
        num_beams: 1,
        penalty_alpha: 0,
        length_penalty: 1,
        early_stopping: false,
        mirostat_mode: 0,
        mirostat_tau: 5,
        mirostat_eta: 0.1,
        seed: -1,
        add_bos_token: true,
        truncation_length: 2048,
        ban_eos_token: false,
        skip_special_tokens: true,
        stopping_strings: ["\n### Assistant:", "\n### Human:"] // I added these, because it seems like, all AI love Assistant and Human
    };

    try {
        console.log(">> SENDING API REQUEST")
        const response = await axios.post(URI, request);

        if (response.status === 200) {
            // Uncommend below, to show full API response
            //console.log("RESPONSE: ", response)

            // The "empty" string that now contains the AI's response
            const result = response.data.results[response.data.results.length - 1].history;
            console.log("Received AI answer: " + result.visible[result.visible.length - 1][1])

            // Set the empty string, that we first pushed for the AI to use for generating it's prompt, to the now received answer, in ["internal"] and ["visible"]
            history.internal[history.internal.length - 1][1] = result.visible[result.visible.length - 1][1]
            history.visible[history.internal.length - 1][1] = result.visible[result.visible.length - 1][1]

            // Uncomment below, to show full history
            //console.log("> NEW HISTORY:\n", history)

            return result.visible[result.visible.length - 1][1]

        } else {
            console.log(` ERROR IN RESPONSE:\n\n\n${response}`)
        }
    } catch (error) {
        console.error('ERROR WHEN SENDING API REQUEST:', error);
    }
}

// Javascriptpart to read the userinput in this example
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Loop 10x for a small conversation with the AI
for (i = 0; i < 10; i++) {
    // Call rl to read userinput
    rl.question(`[${i}] Your Input: `, (input) => {
        console.log('Human:', input);

        // Push a new conversation to the history, but with an empty string (or containing the start of AI reply) for the AI to generate on.
        history.internal.push([input, ""])
        history.visible.push([input, ""])

        console.log

        // Send API request with history.
        run(input, history)
        rl.close();
    });
}
```
